### PR TITLE
pageserver: do vectored read on each dio-aligned section once

### LINKED ISSUE #8719: vectored read: only read each stx_dio_offset_align/stx_dio_mem_align-sized section once
Currently, VectoredReadPlanner issues separate reads as soon as there is a gap between the required reads.

E.g., if we have a reads `[0,20)`, `[20, 30)` and `[50,60)` the result will be two read system calls for `[0,30)` and `[50,60)`.

With direct IO,
* read system calls will have higher latency
* we have  `stx_dio_offset_align` / `stx_dio_mem_align` alignment requirements

So, if `[0,30)` and `[50,60)` fall within the same `stx_dio_offset_align`/`stx_dio_mem_align` section of the file, we should only be issuing one read for these two, using a short-lived IO buffer, and reference or memcpy the data we're actually interested in.

# DoD

Mode for `VectoredReadPlanner` or a wrapper around it that batches reads together into  `stx_dio_offset_align`/`stx_dio_mem_align`-sized reads.

The `max_read_size` still needs to be respected.

### PR DESCRIPTION
**Pull request by yliang412**
Part of #8130, closes #8719.

## Problem

Currently, vectored blob io only coalesce blocks if they are immediately adjacent to each other. When we switch to Direct IO, we need a way to coalesce blobs that are within the dio-aligned boundary but has gap between them.

## Summary of changes

- Introduces a `VectoredReadCoalesceMode` for `VectoredReadPlanner` and `StreamingVectoredReadPlanner` which has two modes: 
  - `AdjacentOnly` (current implementation)
  - `Chunked(<alignment requirement>)`
- New `ChunkedVectorBuilder` that considers batching `dio-align`-sized read, the start and end of the vectored read will respect `stx_dio_offset_align` / `stx_dio_mem_align` (`vectored_read.start` and `vectored_read.blobs_at.first().start_offset` will be two different value).
- Since we break the assumption that blobs within single `VectoredRead` are next to each other (implicit end offset), we start to store blob end offsets in the `VectoredRead`.
- Adapted existing tests to run in both `VectoredReadCoalesceMode`.
- The io alignment can also be live configured at runtime.

## Testing
See https://github.com/neondatabase/neon/pull/8779 for a matrix build of the regression test with alignment requirement = [1, 512].

## Performance

[Benchmark Results](https://www.notion.so/neondatabase/Benchmark-chunked-vectored-read-2024-08-26-0d8db0097df34e3797acad2b3804bcf6?pvs=4)

**TLDR:** No significant difference between using different chunk sizes.

## Rollout

- The adjacent-only merge is enabled by default (`io_buffer_alignment=0`).
- Run Rust unittest running with alignment requirement = [0, 1, 512].

We will test the new chunked vectored read code path in pre-prod later this week after release.

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: pageserver/src/tenant/vectored_blob_io.rs [hunk 1]
```diff
@@ -76,14 +76,107 @@ pub(crate) enum VectoredReadExtended {
...
+        if align == 1 {
+            VectoredReadCoalesceMode::AdjacentOnly
+        } else {
+            VectoredReadCoalesceMode::Chunked(align)
+        }
+    }
+}
+
+pub(crate) enum VectoredReadBuilder {
+    Unconstrained(VectoredReadBuilderInner),
```
Comment (VladLazar) [inline pageserver/src/tenant/vectored_blob_io.rs:100]: nit: not sure what `Unconstrained` means here. It can be constrained with `max_read_size`.

**Review by VladLazar — COMMENTED** [review]
Generally looks good.

---

Are we supporting these two modes to allow for a staged cut-over to dio?

File: pageserver/src/virtual_file.rs [hunk 2]
```diff
@@ -1155,6 +1158,47 @@ fn get_open_files() -> &'static OpenFiles {
...
+    } else {
+        Err(align)
+    }
+}
+
+pub(crate) fn get_io_buffer_alignment() -> usize {
+    let align = IO_BUFFER_ALIGNMENT.load(std::sync::atomic::Ordering::Relaxed);
+
+    if cfg!(test) {
+        let env_var_name = "NEON_PAGESERVER_UNIT_TEST_IO_BUFFER_ALIGNMENT";
```
Comment (VladLazar) [inline pageserver/src/virtual_file.rs:1183]: It would be better to somehow run both flavours of unit tests by default. I don't like the fact that these unit tests won't run in CI (too easy to break unless you know of the magic env var).

File: pageserver/src/tenant/vectored_blob_io.rs [hunk 3]
```diff
@@ -144,6 +237,102 @@ impl VectoredReadBuilder {
...
+    ///
+    /// - The start offset matches with the current end of the vectored read
+    /// - The end block number matches the current end block number.
+    ///
+    /// The resulting size also must be below the max read size.
+    pub(crate) fn extend(&mut self, start: u64, end: u64, meta: BlobMeta) -> VectoredReadExtended {
+        tracing::trace!(start, end, "trying to extend");
+        let start_blk_no = start as usize / self.chunk_size;
+        let end_blk_no = div_round_up(end as usize, self.chunk_size);
+        if (self.last_end_offset() == start || self.end_blk_no == start_blk_no + 1) && {
```
[the remark this answers is not in the document]
Comment (yliang412) [reply pageserver/src/tenant/vectored_blob_io.rs:302]:
```
2 blocks of size 100
         100    199
|------|  |------|
0      99
   A.        B.
```

The `end_blk_no` is exclusive, which means that we have (start_blk_no=0, end_blk_no=1) for both blocks. In this case, the condition `A.end_blk_no == B.start_blk_no + 1` is satisfied.

I think `end_blk_no` is confusing. I will find a better name.

File: pageserver/src/virtual_file.rs (hunk 2 above)
Comment (yliang412) [reply pageserver/src/virtual_file.rs:1183]: I added the following in the CI matrix run branch for the rust unit tests, which we could add to the mergable PR.

https://github.com/neondatabase/neon/blob/c2dcc1585d9de7f514af6339a64c49b9c5311a74/.github/workflows/_build-and-test-locally.yml#L212-L216

Running the regression tests against both alignments in CI is too costly though.

**Comment by yliang412** [comment]
> Are we supporting these two modes to allow for a staged cut-over to dio?

Currently doing perf testing to see if this is mergable without the actual `O_DIRECT` changes. I will report back findings once I'm done.

File: pageserver/src/tenant/vectored_blob_io.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (VladLazar) [reply pageserver/src/tenant/vectored_blob_io.rs:302]: > The end_blk_no is exclusive

Makes sense.

> which means that we have (start_blk_no=0, end_blk_no=1) for both blocks

I don't see this, but you're right they would be merged.
For A: `start_blk_no = 10 / 100 = 0` and `end_block_no = (119 + 99) / 100 = 2`.
For B: `start_blk_no = 120 / 100 = 1` and `end_block_no = (160 + 99) / 100 = 2`.
 
---

Let's take a different example.

```
0     99  100   199 200   299
|------|  |------| |------|
   a          b      c d
```

Blob [a, b]: `start_blk_no = 0` `end_blk_no = 2`
Blob [c, d]: `start_blk_no = 2` `end_blk_no = 3`

In this case it would be desirable to merge but we don't.

[the remark this answers is not in the document]
Comment (yliang412) [reply pageserver/src/tenant/vectored_blob_io.rs]: 15fa58c02852f7bf652a2f0ae380c8857a10f3b4 should take care of this.

File: pageserver/src/tenant/vectored_blob_io.rs [hunk 4]
```diff
@@ -144,6 +237,102 @@ impl VectoredReadBuilder {
...
+            if let Some(max_read_size) = self.max_read_size {
+                let coalesced_size = (end_blk_no - self.start_blk_no) * self.chunk_size;
+                coalesced_size <= max_read_size
+            } else {
+                true
+            }
```
[the remark this answers is not in the document]
Comment (yliang412) [reply pageserver/src/tenant/vectored_blob_io.rs:301]: Part of 15fa58c02852f7bf652a2f0ae380c8857a10f3b4

File: pageserver/src/tenant/vectored_blob_io.rs [hunk 5]
```diff
@@ -76,14 +76,107 @@ pub(crate) enum VectoredReadExtended {
...
+            VectoredReadBuilder::Chunked(builder) => builder.size(),
+        }
+    }
+}
+
+pub(crate) struct VectoredReadBuilderInner {
     start: u64,
     end: u64,
-    blobs_at: VecMap<u64, BlobMeta>,
+    blobs_at: VecMap<u64, (u64, BlobMeta)>,
```
[the remark this answers is not in the document]
Comment (yliang412) [reply pageserver/src/tenant/vectored_blob_io.rs:183]: 722d36f7a86befa69a2f5c071ffbce41c8bebc06

File: pageserver/src/tenant/vectored_blob_io.rs [hunk 6]
```diff
@@ -310,10 +504,11 @@ impl<'a> VectoredBlobReader<'a> {
...
-        let start_offset = blobs_at.first().expect("VectoredRead is never empty").0;
 
-        let mut metas = Vec::with_capacity(blobs_at.len());
+        // Note: read.start != blobs_at.first().start
+        let start_offset = read.start;
 
+        let mut metas = Vec::with_capacity(blobs_at.len());
         // Blobs in `read` only provide their starting offset. The end offset
         // of a blob is implicit: the start of the next blob if one exists
         // or the end of the read.
```
[the remark this answers is not in the document]
Comment (yliang412) [reply pageserver/src/tenant/vectored_blob_io.rs:529]: Correct, we do still need to special case for the last blob though, afc827acdfd5ef5e262aae2a5ebf0406c85e0686.

File: pageserver/src/tenant/vectored_blob_io.rs [hunk 7]
```diff
@@ -76,14 +76,115 @@ pub(crate) enum VectoredReadExtended {
...
+            VectoredReadCoalesceMode::Chunked(chunk_size) => {
+                Self::Chunked(ChunkedVectoredReadBuilderInner::new(
+                    start_offset,
+                    end_offset,
+                    meta,
+                    None,
+                    chunk_size,
+                ))
+            }
+        }
```
Comment (problame) [inline pageserver/src/tenant/vectored_blob_io.rs:152]: Bit inelegant & bloaty; why not one `new_impl` to which both `new` and `new_streaming` forward?

The only difference is the `Some()` vs `None` in the bodies.
That's the important difference you want the reader to be able to spot easily.

File: pageserver/src/tenant/vectored_blob_io.rs [hunk 8]
```diff
@@ -76,14 +76,115 @@ pub(crate) enum VectoredReadExtended {
...
+        } else {
+            VectoredReadCoalesceMode::Chunked(align)
+        }
+    }
+}
+
+pub(crate) enum VectoredReadBuilder {
+    Adajacent(AdjacentVectoredReadBuilderInner),
+    Chunked(ChunkedVectoredReadBuilderInner),
+}
```
Comment (problame) [inline pageserver/src/tenant/vectored_blob_io.rs:102]: Why the `Inner` suffix on these types? You already have the `Adjacent|Chunked` prefix.

Also, typo:


```suggestion
pub(crate) enum VectoredReadBuilder {
    Adjacent(AdjacentVectoredReadBuilderInner),
    Chunked(ChunkedVectoredReadBuilderInner),
}
```

File: pageserver/src/tenant/vectored_blob_io.rs [hunk 9]
```diff
@@ -303,34 +512,43 @@ impl<'a> VectoredBlobReader<'a> {
...
             .file
             .read_exact_at(buf.slice(0..read.size()), read.start, ctx)
             .await?
             .into_inner();
 
         let blobs_at = read.blobs_at.as_slice();
-        let start_offset = blobs_at.first().expect("VectoredRead is never empty").0;
 
-        let mut metas = Vec::with_capacity(blobs_at.len());
+        // Note: read.start != blobs_at.first().start
```
Comment (problame) [inline pageserver/src/tenant/vectored_blob_io.rs:535]: Don't think this comment is helpful down the line.

File: pageserver/src/tenant/vectored_blob_io.rs [hunk 10]
```diff
@@ -353,10 +571,13 @@ impl<'a> VectoredBlobReader<'a> {
...
             };
 
-            let start_raw = offset_in_buf + size_length;
-            let end_raw = match next {
-                Some((next_blob_start_offset, _)) => next_blob_start_offset - start_offset,
-                None => start_raw + blob_size,
+            let start_raw = blob_start_in_buf + size_length;
+            let end_raw = {
+                if *blob_start == last_blob_start {
+                    start_raw + blob_size
```
Comment (problame) [inline pageserver/src/tenant/vectored_blob_io.rs:577]: Why is this necessary? Is the `blob_end` not valid in the last element of `blobs_at`?

It seems to me this is a sort of pattern-matched port of Some / None logic that we had with the `.zip()` before this PR.

**Review by problame — COMMENTED** [review]
This review pertains to code that is taken by the `Adjacent` path, i.e., this review is to ensure we're not regressing anything.

File: pageserver/src/tenant/vectored_blob_io.rs [hunk 11]
```diff
@@ -144,6 +247,107 @@ impl VectoredReadBuilder {
...
+        /// Note: This automatically handles the case where two blocks are adjacent to each other,
+        /// whether they starts on chunk size boundary or not.
+        fn is_adjacent_chunk_read(first_end_blk_no: usize, second_start_blk_no: usize) -> bool {
+            // 1. first.end & second.start are in the same block
+            first_end_blk_no == second_start_blk_no + 1 ||
+            // 2. first.end ends one block before second.start
+            first_end_blk_no == second_start_blk_no
+        }
+
+        if is_adjacent_chunk_read(self.end_blk_no, start_blk_no) && not_limited_by_max_read_size() {
```
Comment (problame) [inline pageserver/src/tenant/vectored_blob_io.rs:324]: Style: why is this an function? It's called exactly once.

Why not

```
let is_adjacent_chunk_read = { ...};

if is_adjacent_chunk_read && not_limited_by_max_read_size { ... }
```

File: pageserver/src/tenant/vectored_blob_io.rs [hunk 12]
```diff
@@ -144,6 +247,107 @@ impl VectoredReadBuilder {
...
+        let end_blk_no = div_round_up(end as usize, self.chunk_size);
+
+        let not_limited_by_max_read_size = || {
+            if let Some(max_read_size) = self.max_read_size {
+                let coalesced_size = (end_blk_no - self.start_blk_no) * self.chunk_size;
+                coalesced_size <= max_read_size
+            } else {
+                true
+            }
+        };
```
Comment (problame) [inline pageserver/src/tenant/vectored_blob_io.rs:311]: style / weird: why not simply 

```
let not_limited_by_max_read_size = { ... }
```

I would naively (?) trust the compiler to rearrange the control flow.

(There is another occurency of this pattern in `AdjacentVectoredReadBuilderInner`)

**Review by problame — COMMENTED** [review]
Reviewed the ChunkedVectoredReadBuilderInner.

Seems correct, some style comments.

File: pageserver/src/tenant/vectored_blob_io.rs (hunk 10 above)
Comment (yliang412) [reply pageserver/src/tenant/vectored_blob_io.rs:577]: > Why is this necessary? Is the blob_end not valid in the last element of blobs_at?

`blob_end` is indeed not valid for the last element of `blobs_at`. This is used to handle the following case, where the end offset is set to `data_offset`.

https://github.com/neondatabase/neon/blob/84f770b26516b45516b2429b6b7ba9cc60525ee6/pageserver/src/tenant/storage_layer/delta_layer.rs#L968-L971

We obtain this `data_offset` from https://github.com/neondatabase/neon/blob/84f770b26516b45516b2429b6b7ba9cc60525ee6/pageserver/src/tenant/storage_layer/delta_layer.rs#L1425-L1438

Note L1427: 
> The file structure is summary, values, and index. We often need this for the size of last blob.

I would this is not always the case if the last blob ends earlier than the page boundary.

If we turn off the special casing, then tests will fail:

```console
--- STDERR:              pageserver tenant::storage_layer::delta_layer::test::test_delta_layer_vectored_read_end_to_end ---
thread 'tenant::storage_layer::delta_layer::test::test_delta_layer_vectored_read_end_to_end' panicked at pageserver/src/tenant/vectored_blob_io.rs:576:13:
assertion `left == right` failed
  left: 5165
 right: 100
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```




Options I see:
1. Always use parsed blob_size + remove assertions on L582. 
2. keep doing the special casing.

https://github.com/neondatabase/neon/blob/84f770b26516b45516b2429b6b7ba9cc60525ee6/pageserver/src/tenant/vectored_blob_io.rs#L576-L582

Comment (problame) [reply pageserver/src/tenant/vectored_blob_io.rs:577]: cc @VladLazar , so, the summary block is actually padded up to the next page boundary

File: pageserver/src/tenant/vectored_blob_io.rs (hunk 7 above)
Comment (yliang412) [reply pageserver/src/tenant/vectored_blob_io.rs]: Thanks for the advice! 124c97fe6847dc06ecd41e1a387bb0071aa93c96

File: pageserver/src/tenant/vectored_blob_io.rs (hunk 8 above)
Comment (yliang412) [reply pageserver/src/tenant/vectored_blob_io.rs]: d0fd9b5d4a496d6c40687bfd6ba114ed5213bf86

File: pageserver/src/tenant/vectored_blob_io.rs (hunk 9 above)
Comment (yliang412) [reply pageserver/src/tenant/vectored_blob_io.rs]: removed in fe2393c8c3a4ae37890f87f9d767489f4acf51af

File: pageserver/src/tenant/vectored_blob_io.rs (hunk 11 above)
Comment (yliang412) [reply pageserver/src/tenant/vectored_blob_io.rs]: Agreed. 9213dc31e60af84da35531598b462fb9db1cc9b5

File: pageserver/src/tenant/vectored_blob_io.rs (hunk 12 above)
Comment (yliang412) [reply pageserver/src/tenant/vectored_blob_io.rs]: fixed both occurences in e5bbece5669dedd32d4b9ce85c1603bbeb64ffbd

File: pageserver/src/tenant/vectored_blob_io.rs (hunk 10 above)
Comment (yliang412) [reply pageserver/src/tenant/vectored_blob_io.rs]: refactored in 70dc350ffd107acc07e4b19e7d1a292236434e0d. We will directly use start offset and parsed blob_size to determine the end offset. As a result, 

1. stopped storing `end_offset` for each blob.
2. no special casing for last blob in `read_blobs`
3. removed `assert_eq!(end_raw - start_raw, blob_size)` assertion.

cc. @VladLazar, I think we can trust the parsed blob size?

File: pageserver/src/tenant/vectored_blob_io.rs [hunk 13]
```diff
@@ -61,7 +61,7 @@ pub struct VectoredRead {
     pub start: u64,
     pub end: u64,
     /// Starting offsets and metadata for each blob in this read
-    pub blobs_at: VecMap<u64, BlobMeta>,
+    pub blobs_at: VecMap<u64, (u64, BlobMeta)>,
```
[the remark this answers is not in the document]
Comment (yliang412) [reply pageserver/src/tenant/vectored_blob_io.rs]: We no longer keep `end_offset` for the blobs as discussed in https://github.com/neondatabase/neon/pull/8763#discussion_r1730426970.

File: pageserver/src/tenant/vectored_blob_io.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (yliang412) [reply pageserver/src/tenant/vectored_blob_io.rs:532]: See https://github.com/neondatabase/neon/pull/8763#discussion_r1730426970

**Review by problame — APPROVED** [review]
